### PR TITLE
Add a new value_init test, change implementation to not depend on TypeTraits

### PR DESCRIPTION
**Pull request by pdimov**
As suggested in https://github.com/boostorg/core/issues/76.

### CODE REVIEW

**Comment by Lastique** [comment]
Though this will likely work, isn't using `memset` on a half-constructed non-POD object UB?

**Comment by pdimov** [comment]
Interesting question. I'm not sure what would make it UB.

**Comment by Lastique** [comment]
Basically, object representation of non-trivial and non-standard-layout types is implementation-defined, so modifying it with `memset` and friends is UB. [intro.object]/8 doesn't even guarantee that such objects are contiguous.

In practice, though, `zero_init` and `initialized` will likely have the natural contiguous layout with no hidden members like vtable pointer etc.

**Comment by Lastique** [comment]
You could probably `memset` `data_` in `zero_init` constructor as that would be zeroing the storage of that specific member before it is constructed.

**Comment by Kojoley** [comment]
Writing anything to anything is not UB, reading after it may. One may be concerned about virtual classes, but the standard is worded in a way that the virtual pointer is actually written after user constructor is called. UBSAN (of both Clang and GCC) is also quite about the trick.

**Comment by Kojoley** [comment]
@zygoloid could you take a look and judge whether the code contains UB, please?

**Comment by Lastique** [comment]
> the standard is worded in a way that the virtual pointer is actually written after user constructor is called

Given that the virtual function is called in the class that is being constructed/destructed, I'd say the vtable pointer must be written before that construction/destruction begins.

**Comment by Kojoley** [comment]
Anyway it should happen only after `zero_init` constructor is done. Again, UBSAN is fine with:
```cpp
#include <cstring>

struct foo
{
    foo() : i(get() + 42) { }
    virtual ~foo() {}
    virtual int get() const { return 42; }
    int i;
};

struct bar : virtual foo
{
    bar() {}
    int get() const override { return i; }
};

struct zero_init
{
    zero_init( void * p, std::size_t n )
    {
        std::memset( p, 0, n );
    }
};

template<class T> class initialized: private zero_init
{
private:

    T t_;

public:

    initialized(): zero_init( this, sizeof(*this) ), t_()
    {
    }

    T const& get() const noexcept
    {
        return t_;
    }
};


int main()
{
    initialized<bar> i;
    return i.get().get();
}
```

**Comment by pdimov** [comment]
> Given that the virtual function is called in the class that is being constructed/destructed

I don't think they are.

**Comment by pdimov** [comment]
> You could probably `memset` `data_` in `zero_init` constructor as that would be zeroing the storage of that specific member before it is constructed.

This is a good idea, thanks.

**Comment by Lastique** [comment]
> I don't think they are.

They are, see [[class.cdtor]/4](http://eel.is/c++draft/class.cdtor#4).

**Comment by Kojoley** [comment]
> > You could probably memset data_ in zero_init constructor as that would be zeroing the storage of that specific member before it is constructed.
>
> This is a good idea, thanks.

It gives no additional safety but requires `addressof`

**Comment by Lastique** [comment]
@Kojoley I think you may have misunderstood my point. The problem is not the representation of `bar` in your example. It's that of `initialized` and `zero_init`. The standard doesn't guarantee anything about them and in particular allows the implementation to add hidden members in them or do any kind of funny stuff.

**Comment by pdimov** [comment]
> They are, see [[class.cdtor]/4](http://eel.is/c++draft/class.cdtor#4).

In the constructor of the derived class, virtual calls resolve to that derived class. In the constructor of the base class (`zero_init` in our case), virtual calls resolve to that base class. https://godbolt.org/z/a21ZTQ

It follows that the vptr is initialized to point at the derived class vtable after the base class constructor completes, i.e. after the memset.

Anyway, I've updated the PR to memset `data_` instead of `*this`, which is more correct in either case.

**Comment by Kojoley** [comment]
> The standard doesn't guarantee anything about them and in particular allows the implementation to add hidden members in them or do any kind of funny stuff.

I do not understand why does it matter, invoking constructor on uninitialized memory would had the same problem in that case.

**Comment by Lastique** [comment]
> In the constructor of the derived class, virtual calls resolve to that derived class. In the constructor of the base class (`zero_init` in our case), virtual calls resolve to that base class.

Yes, exactly.

The `memset` overwrites `zero_init` representation as well as `initialized`, which means if the implementation had some hidden member `zero_init` it would have been corrupted. Whether that would be a problem or not would depend on how that member would have been used by the implementation. For example, it could have used it before calling `initialized` constructor.

> Anyway, I've updated the PR to memset data_ instead of *this, which is more correct in either case.

Great, thanks!

**Comment by pdimov** [comment]
There's still one potentially problematic case,
```
struct X
{
    int a;
    char b;
};

struct Y: value_initialized<X>
{
    char c = 42;
};
```
where the `memset` may overwrite `c` which could have been laid out in the padding of `X` and initialized before calling the constructor of `zero_init`. I'm not sure if that's a problem in practice though.

**Comment by Lastique** [comment]
I think this should not happen if `X` is stored as a member in `value_initialized` (which it is). `c` can only be stored in the tail padding if `X` is potentially overlapping (i.e. a base class or marked with `[[no_unique_address]]`).

File: test/value_init_test3.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// Copyright 2020 Peter Dimov
+// Distributed under the Boost Software License, Version 1.0.
+// https://www.boost.org/LICENSE_1_0.txt)
+
+#include <boost/utility/value_init.hpp>
+#include <boost/core/lightweight_test.hpp>
+#include <boost/config.hpp>
+#include <boost/config/pragma_message.hpp>
+
+#if __cplusplus >= 201103L || ( defined(BOOST_MSVC) && BOOST_MSVC >= 1900 )
```
Comment (Lastique) [inline test/value_init_test3.cpp:10]: Doesn't `BOOST_NO_CXX14_AGGREGATE_NSDMI` cover this?

Comment (pdimov) [reply test/value_init_test3.cpp:10]: No.

Comment (pdimov) [reply test/value_init_test3.cpp:10]: BOOST_NO_CXX14_AGGREGATE_NSDMI is about the C++14 feature, the equivalent of __cpp_aggregate_nsdmi. What I need here is the nonexistent macro about the C++11 feature, the equivalent of __cpp_nsdmi. No idea why we have the former in Config, but not the latter.

Comment (Lastique) [reply test/value_init_test3.cpp:10]: Ok, thanks. Maybe we should add one.

**Pull request merged**